design.md

```diff
@@ -0,0 +1,138 @@
+  - Remove a plugin by its ID.
+- Each plugin has a unique ID and a Jar file.
+- The kernel must be able to download simulator plugins.
+  - We can ship the default simulator plugin with the IntelliJ plugin, but there must be a way to download any of them because of the [usability requirement](#usability).
+
+#### UI Interaction
+
+- ! RFC: The kernel must be able to notify the UI of multiple tasks happening in parallel. For example, running a script could cause some repos to get cloned and a toolchain to get downloaded in parallel. As a user, I want to see both of these operations as tasks in IntelliJ happening in parallel. As a general rule, a script can start multiple tasks.
+- ! RFC: The kernel needs a way to ask the UI for confirmation with a custom confirmation message. For example, if the kernel needs to flash the device with a new program, by default it should ask the user for confirmation. There also needs to be some local environment configuration option that bypasses this so that the confirmation results is either always allowed or denied (for headless operation). For example, you may want to always allow the kernel to flash the device during development or during some competitions (as a user, I would rather have a delayed start to my match rather than having the robot not work at all; it depends on the competition). As another example, you may want to always stop the kernel from flashing the device if you have released a product and you expect all the products to work the same for all your users and would rather handle these issues with your own support system.
+
+#### Timeout
+
+If no script is running and no gRPC interaction has occurred for some time period, then the kernel should exit.
+
+### Headless Execution
+
+- The kernel must not include any references to JavaFX so that it can run in a headless environment.
+
+### Data Logging
+
+- Every connected device resource must be automatically available as a data source. The resources should be loaded from the robot config file.
+  - Sensors return their sensor value (e.g., potentiometer -> voltage, encoder -> ticks, IMU -> values of axes).
+  - Actuators return their control signal (e.g., servo -> angle, solenoid -> digital state, motor -> voltage). These are the values the user's program commands, not the measured value (which would be a sensor).
+- A script may add its own data sources via *variables*.
+  - A variable is addressed by name (a String).
+  - State changes of variables are tracked such that the current state of a variable can be read from another thread in parallel.
+- Data types that can be logged:
+  - Numerical types (e.g., int, long, float, double, byte, char, short, boolean).
+  - Strings.
+  - Frame transforms.
+  - Images (e.g., frames from a camera feed).
+  - Arbitrary bytes.
```

Comment:
What about aggregate data structures composed of these types? (Structured types might be hard because of the way Java works, but lists and/or maps should be feasible). Also, can variables be read over gRPC? (That might be useful for diagnostic purposes).

Reply:
I have added some aggregate data types.

> Also, can variables be read over gRPC?

I still don't want to support this, at least not exclusively. I can see a use case for wanting the whole of bowler-display to work with a remote kernel. Maybe we should support that instead. The use case I have in mind is:
1. You are at a competition and you are working on the robot.
2. You connect to the robot (so the kernel is running headlessly on the robot's embedded computer and you are interfacing with it as a remote kernel).
3. You want to run a script to test something and look at the data logging or the 3D view.